### ISSUE TITLE
Audit and update `google/golang/org/grpc` instrumentation to comply with RPC semantic conventions

### ISSUE THREAD

**Issue by MrAlias** [open-telemetry/opentelemetry-go-instrumentation#172]
https://github.com/open-telemetry/opentelemetry-specification/blob/84a5fa3d1ef42caae8f99f1b431504d7cdb9fffe/specification/trace/semantic_conventions/rpc.md

**Comment by damemi** [comment]
Latest RPC conventions: https://github.com/open-telemetry/semantic-conventions/blob/02ecf0c71e9fa74d09d81c48e04a132db2b7060b/docs/rpc/rpc-spans.md and specifically the gRPC conventions (which extend and override RPC): https://github.com/open-telemetry/semantic-conventions/blob/02ecf0c71e9fa74d09d81c48e04a132db2b7060b/docs/rpc/grpc.md

> ## Common remote procedure call conventions

> For outgoing requests, the `SpanKind` MUST be set to `CLIENT` and for incoming requests to `SERVER`.

Correct - [client](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/9019008bbc20036eb6591372799d773eb89c8428/internal/pkg/instrumentation/bpf/google.golang.org/grpc/client/probe.go#L34)
Correct - [server](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/9019008bbc20036eb6591372799d773eb89c8428/internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/probe.go#L33)

>Remote procedure calls can only be represented with these semantic conventions, when the names of the called service and method are known and available.

>### Span name
>
>The _span name_ MUST be the full RPC method name formatted as:
>
>```
>$package.$service/$method
>```
>
>(where $service MUST NOT contain dots and $method MUST NOT contain slashes)

We don’t seem to currently fit this in our example, see the [gRPC test spans](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/9019008bbc20036eb6591372799d773eb89c8428/internal/test/e2e/grpc/traces.json#L111), which for example set `name` to `/helloworld.Greeter/SayHello` which has a leading `/`. However, this is coming directly from the upstream [example code](https://github.com/grpc/grpc-go/blob/4ed81800b06dd7c587dd42830a780d2dad6812f4/examples/helloworld/helloworld/helloworld_grpc.pb.go#L36) which sets the full method name to that exact string.

We are grabbing this value directly from the `grpc.Invoke()` arguments ([client](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/9019008bbc20036eb6591372799d773eb89c8428/internal/pkg/instrumentation/bpf/google.golang.org/grpc/client/bpf/probe.bpf.c#L89), [server](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/9019008bbc20036eb6591372799d773eb89c8428/internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/probe.go#L44)) so it’s possible that this is just one case where `/helloworld` *is* `$service` as set by the go gRPC library’s generated code.

> If there is no package name or if it is unknown, the `$package.` part (including the period) is omitted.

> ### Service name

> On the server process receiving and handling the remote procedure call, the service name provided in `rpc.service` does not necessarily have to match the [`service.name`][] resource attribute.
>
> One process can expose multiple RPC endpoints and thus have multiple RPC service names. From a deployment perspective, as expressed by the `service.*` resource attributes, it will be treated as one deployed service with one `service.name`.
>
>Likewise, on clients sending RPC requests to a server, the service name provided in `rpc.service` does not have to match the [`peer.service`][] span attribute.

>As an example, given a process deployed as `QuoteService`, this would be the name that goes into the `service.name` resource attribute which applies to the entire process.
>This process could expose two RPC endpoints, one called `CurrencyQuotes` (= `rpc.service`) with a method called `getMeanRate` (= `rpc.method`) and the other endpoint called `StockQuotes`  (= `rpc.service`) with two methods `getCurrentBid` and `getLastClose` (= `rpc.method`).
>In this example, spans representing client request should have their `peer.service` attribute set to `QuoteService` as well to match the server's `service.name` resource attribute.
Generally, a user SHOULD NOT set `peer.service` to a fully qualified RPC service name.

>### Client attributes

The following attributes are required:

* `rpc.system` ([set](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/dfd43b38c0d199cac270871e41816ceef36f8547/internal/pkg/instrumentation/bpf/google.golang.org/grpc/client/probe.go#L109))
* `server.address` ([set](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/dfd43b38c0d199cac270871e41816ceef36f8547/internal/pkg/instrumentation/bpf/google.golang.org/grpc/client/probe.go#L111))
* `server.port` (conditionally, “if the port is supported by the network transport used for communication”) (not set)

> `rpc.system` has the following list of well-known values. If one of them applies, then the respective value MUST be used; otherwise, a custom value MAY be used.

Only one that applies to us is `grpc`, which we set in both [server](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/9019008bbc20036eb6591372799d773eb89c8428/internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/probe.go#L133) and [client](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/dfd43b38c0d199cac270871e41816ceef36f8547/internal/pkg/instrumentation/bpf/google.golang.org/grpc/client/probe.go#L109) 

> ### Server attributes

The following are required:

* `rpc.system` ([set](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/9019008bbc20036eb6591372799d773eb89c8428/internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/probe.go#L133))
* `server.address` (not set)
* `server.port` (conditionally, “if the port is supported by the network transport used for communication”) (not set)

## gRPC specific conventions

See https://github.com/open-telemetry/semantic-conventions/blob/02ecf0c71e9fa74d09d81c48e04a132db2b7060b/docs/rpc/grpc.md

The following attributes are required on both server and client:

* `rpc.grpc.status_code` (not set)

# Updates needed

- [ ] Confirm `service.name` is inherited from the Go gRPC library, which in theory should be compliant with the standard by default
- [ ] Set `server.port` in client
- [ ] Set `server.address` in server
- [ ] Set `server.port` in server
- [ ] Set `rpc.grpc.status_code` in client and server

**Comment by damemi** [comment]
For `rcp.grpc.status_code`, I think if we update our return probe instrumentation for `Invoke` to read the returned error object, that should get us what we need. gRPC seems to return these as a [`Status`](https://github.com/grpc/grpc-go/blob/2da976983bbb33feb3e25b7daaa8f60b9769adb5/internal/status/status.go#L43) struct with another embedded [`*spb.Status](https://github.com/googleapis/go-genproto/blob/2c9e96a0b5d43678edfc0fa92f26dac269fd3374/googleapis/rpc/status/status.pb.go#L46) struct inside it. From there, the code is just an int which we can map to an actual code in Go. So we should just be able to generate the appropriate offsets and read the int value from those.

@RonFed do we have any way to read the return values for an instrumented function?

**Comment by RonFed** [comment]
@damemi The returned `error` is an interface. Assuming we know the concrete type of the error returned, we can add that type and its relevant fields to `offsetgen` and read them in the return probe. If the concrete type is not always the same it will be more challenging since we'll have to understand what struct is being returned in the error.

**Comment by damemi** [comment]
@RonFed yeah, that's what I'm saying is that gRPC should always return the error's concrete type as a [`Status`](https://github.com/grpc/grpc-go/blob/2da976983bbb33feb3e25b7daaa8f60b9769adb5/internal/status/status.go#L43) struct. Because I think it's expected that you can use the [gRPC status package](https://pkg.go.dev/google.golang.org/grpc/status) to parse this response code, I'm pretty sure that's a safe assumption.

Do we have any examples of reading return values? I know we read the method arguments

**Comment by RonFed** [comment]
@damemi Yes,

Here we are reading the returned pointer from a register:
https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/68865cc10e60bfeeed351827a596c34a993093d4/internal/pkg/instrumentation/bpf/go.opentelemetry.io/otel/traceglobal/bpf/probe.bpf.c#L388-L389

Here we are reading the returned value of the stack:
https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/68865cc10e60bfeeed351827a596c34a993093d4/internal/pkg/instrumentation/bpf/github.com/segmentio/kafka-go/consumer/bpf/probe.bpf.c#L166-L167

In the error case it will be returned as a pointer in a register, we can read the pointer and then the fields.

**Comment by damemi** [comment]
Awesome, thanks! I'll try to follow that and implement it for grpc
